test/s3-benchmarks/src/main/java/software/amazon/awssdk/s3benchmarks/BaseJavaS3ClientBenchmark.java

```diff
@@ -0,0 +1,130 @@
+        this.checksumAlgorithm = config.checksumAlgorithm();
+
+        this.s3Client = S3Client.create();
+
+        long partSizeInMb = Validate.paramNotNull(config.partSizeInMb(), "partSize");
+        long readBufferInMb = Validate.paramNotNull(config.readBufferSizeInMb(), "readBufferSizeInMb");
+        Validate.mutuallyExclusive("cannot use forceCrtHttpClient and connectionAcquisitionTimeoutInSec",
+                                   config.forceCrtHttpClient(), config.connectionAcquisitionTimeoutInSec());
+        this.s3AsyncClient = S3AsyncClient.builder()
+                                          .multipartEnabled(true)
+                                          .multipartConfiguration(c -> c.minimumPartSizeInBytes(partSizeInMb * MB)
+                                                                        .thresholdInBytes(partSizeInMb * 2 * MB)
+                                                                        .apiCallBufferSizeInBytes(readBufferInMb * MB))
+                                          .httpClient(httpClient(config))
+                                          .build();
+    }
+
+    private SdkAsyncHttpClient httpClient(TransferManagerBenchmarkConfig config) {
+        if (config.forceCrtHttpClient()) {
+            logger.info(() -> "Using CRT HTTP client");
+            AwsCrtAsyncHttpClient.Builder builder = AwsCrtAsyncHttpClient.builder();
+            if (config.readBufferSizeInMb() != null) {
+                builder.readBufferSizeInBytes(config.readBufferSizeInMb());
+            }
+            if (config.maxConcurrency() != null) {
+                builder.maxConcurrency(config.maxConcurrency());
+            }
+            return builder.build();
+        }
+        NettyNioAsyncHttpClient.Builder builder = NettyNioAsyncHttpClient.builder();
+        if (config.connectionAcquisitionTimeoutInSec() != null) {
+            Duration connAcqTimeout = Duration.ofSeconds(config.connectionAcquisitionTimeoutInSec());
+            builder.connectionAcquisitionTimeout(connAcqTimeout);
+        }
+        if (config.maxConcurrency() != null) {
+            builder.maxConcurrency(config.maxConcurrency());
+        }
+        return builder.build();
+    }
+
+    protected abstract void sendOneRequest(List<Double> latencies) throws Exception;
+
+    protected abstract long contentLength() throws Exception;
+
+    @Override
+    public void run() {
+        try {
+            warmUp();
+            doRunBenchmark();
+        } catch (Exception e) {
+            logger.error(() -> "Exception occurred", e);
+        } finally {
+            cleanup();
+        }
+    }
+
+    private void cleanup() {
+        s3Client.close();
+        s3AsyncClient.close();
+    }
+
+    private void warmUp() throws Exception {
+        logger.info(() -> "Starting to warm up");
+        for (int i = 0; i < 3; i++) {
+            sendOneRequest(new ArrayList<>());
```

Comment:
This probably isn't enough warming to get everything fully warmed.

Reply:
I used the same warmup logic that is used for CRT Benchmarks to be consistent between the CRT's `S3Client` and the java-based `MultipartS3AsyncClient`. I could bump both to 5?

Reply:
Honestly, I'd expect hundreds to get things properly inlined, JITed, etc. I guess we can leave this as-is for now and wait for JMH, since it understands this much better than all of us.

Reply:
From my previous tests, the startup time doesn't matter that much for tests in this module since we are transferring a large object or a lot of objects, which could take 30 seconds+. But agreed, JMH would probably work better. I don't remember why I didn't go with JMH in the initial version, probably ran into some issue, but it's definitely worth another try!